pallets/multisig-verifier/src/lib.rs

```diff
@@ -0,0 +1,294 @@
+    }
+
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    #[pallet::without_storage_info]
+    pub struct Pallet<T>(_);
+
+    #[pallet::storage]
+    #[pallet::getter(fn get_peer_keys)]
+    pub type PeerKeys<T> =
+        StorageMap<_, Twox64Concat, GenericNetworkId, BTreeSet<ecdsa::Public>, OptionQuery>;
+
+    #[pallet::type_value]
+    pub fn DefaultForThisNetworkId() -> GenericNetworkId {
+        GenericNetworkId::Sub(SubNetworkId::Mainnet)
+    }
+
+    #[pallet::storage]
+    #[pallet::getter(fn this_network_id)]
+    pub type ThisNetworkId<T> = StorageValue<_, GenericNetworkId, ValueQuery, DefaultForThisNetworkId>;
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        NetworkInitialized(GenericNetworkId),
+        VerificationSuccessful(GenericNetworkId),
+        PeerAdded(ecdsa::Public),
+        PeerRemoved(ecdsa::Public),
+
+        // Error events:
+
+        NetworkNotInitialized(GenericNetworkId),
+        /// NetworkId, Required Number, Current Number
+        InvalidNumberOfSignatures(GenericNetworkId, u32, u32),
+        InvalidSignature(GenericNetworkId, H256, ecdsa::Signature),
+        NotTrustedPeerSignature(GenericNetworkId, H256, ecdsa::Signature, ecdsa::Public),
+        NoSuchPeer(ecdsa::Public),
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        InvalidInitParams,
+        NetworkNotInitialized,
+        InvalidNumberOfSignatures,
+        InvalidSignature,
+        NotTrustedPeerSignature,
+        NoSuchPeer,
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+    #[pallet::genesis_config]
+    pub struct GenesisConfig {
+        /// Network id for current network
+        pub network_id: GenericNetworkId,
+    }
+
+    #[cfg(feature = "std")]
+    impl Default for GenesisConfig {
+        fn default() -> Self {
+            Self {
+                network_id: GenericNetworkId::Sub(SubNetworkId::Mainnet),
+            }
+        }
+    }
+
+    #[pallet::genesis_build]
+    impl<T: Config> GenesisBuild<T> for GenesisConfig {
+        fn build(&self) {
+            ThisNetworkId::<T>::put(self.network_id);
+        }
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        #[pallet::call_index(0)]
+        #[pallet::weight(0)]
+        pub fn initialize(
+            origin: OriginFor<T>,
+            network_id: GenericNetworkId,
+            keys: Vec<ecdsa::Public>,
+        ) -> DispatchResultWithPostInfo {
+            ensure_root(origin)?;
+            ensure!(keys.len() > 0, Error::<T>::InvalidInitParams);
+
+            let btree_keys = keys.into_iter().collect();
+            PeerKeys::<T>::set(network_id, Some(btree_keys));
+            Self::deposit_event(Event::NetworkInitialized(network_id));
+            Ok(().into())
+        }
+
+        #[pallet::call_index(1)]
+        #[pallet::weight(0)]
+        pub fn add_peer(
```

Comment:
Also please submit DataSignerCall::AddPeer message back to origin network to finish adding of peer. Same for remove_peer